### PR TITLE
[LSP] Support LSP services associated with LSP server instances (with lifetimes that match).

### PR DESCRIPTION
**Pull request by dibarbet**
Resolves https://github.com/dotnet/roslyn/issues/61002

This PR creates the concept of an `ILspService` and `ILspServiceFactory`.  These can be exported via two different attributes
1.  `ExportStatelessLspService` attribute exports an `ILspService` that must be stateless.  The instance exported is re-used across all LSP server instances throughout the lifetime of the process.
2.  `ExportLspServiceFactory` exports an `ILspServiceFactory`.  The factory instance is shared across all servers, however the `ILspService` creation method is called for each individual server instance.  This allows the service to have an instance per server.  The services exported via the factory are also disposed of on server shutdown.

Each of the attributes specifies 
1.  the concrete type that implements `ILspService` - used to find the service via `GetRequiredService<T>()`
2.  the MEF contract name - used to ensure we only import services for C# when we're in C#, only Xaml when in Xaml, and only JS when in JS to avoid RPS regressions.
3.  An optional server kind parameter - used if an lsp server should only apply to a specific LSP server (for example you want to provide a service only for the C# razor server but not the normal C# server).

The majority of the code changes is just a bunch of boilerplate switching all the request handlers and a few other things to use the lsp service concept.

RPS val - https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/399013

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/LspServices/LspServices.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/LspServices/LspServices.cs:22]: @sharwell I wanted to get your MEF expertise on this to see if this was the best way to do this.  Basically what I want is to be able to create new instances of the `ILspService` for each LSP server instance (could be multiple running at one time), dispose of them on shutdown, and have them be recreated properly if the server shutsdown / restarts.

The best way I could come up with is making all exports for `ILspService` not shared and manually retrieve them on server creation here to force mef to create new ones.  

However this will require me to suppress `#RS0023 // Parts exported with MEFv2 must be marked with 'SharedAttribute'` as these cannot be shared instances.  I think I can write a diagnostic suppressor for that?  But wanted to check with you first if you had a better approach.

Comment (sharwell) [reply src/Features/LanguageServer/Protocol/LspServices/LspServices.cs:22]: Rather than export `ILspService`, you can export `ILspServiceFactory`. This allows the factory to be static, and also allows the factory to cache the service instance if it wants to. See `IWorkspaceServiceFactory` for another example.

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/LspServices/LspServices.cs]: basically did this - there is some extra boiler plate with this approach (e.g. https://github.com/dotnet/roslyn/pull/61266/files#diff-e49357a9ef274059c786a87abe0c688392ac8101a18dc41327d1aaa347c17243R20 ) but the majority are stateless and so don't need to actually use a factory.  

To indicate which is which I have two attributes - one for exporting stateless LSP services, and one for exporting lsp service factories

**Comment by dibarbet** [comment]
running RPS before requesting further reviews - https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6148840&view=results

File: src/Features/LanguageServer/Protocol/Handler/IDocumentChangeTracker.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/IDocumentChangeTracker.cs:14]: moved into its own class, unchanged.

File: src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs:264]: all this was moved unchanged into this file from a previous partial

File: src/Features/LanguageServer/Protocol/Handler/RequestTelemetryLogger.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/RequestTelemetryLogger.cs:33]: all this was moved unchanged into its own file

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs:106]: now this can be initialized in the ctor instead of on first request

File: src/Features/LanguageServer/Protocol/LspServices/LspServices.cs
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/LspServices/LspServices.cs:17]: new interesting code

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLspFactories.cs
Comment (ryanbrandenburg) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLspFactories.cs:15]: Nit: In Razor we try to keep it to 1 file in 1 class (unless the second is private or some such), but definitely keep with whatever the standard is in Roslyn.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLspServiceProvider.cs
Comment (ryanbrandenburg) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptLspServiceProvider.cs:13]: Nit: Personally I'd explicitly define the class that this exports for clarity.

File: src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionResolveHandler.cs
Comment (ryanbrandenburg) [inline src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionResolveHandler.cs:73]: I assume the cache is thread-safe?

File: src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsCache.cs
Comment (ryanbrandenburg) [inline src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsCache.cs:29]: If we return a new cache each time this is called and `CodeActionResolveHandler.cs` doesn't hold it then I assume that the Context object is ensuring that only one of these is created for the whole process?

My paranoia makes me think it might be good to add a quick check here to see that this is only called once if that's the case.

File: src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsHandler.cs
Comment (ryanbrandenburg) [inline src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsHandler.cs:64]: If we get the same codeActionsCache back each time this is called does it maybe make more sense to just hold onto it after the first time instead of constantly getting the service again? I'll admit this is cleaner but I wonder if the overhead of GetRequiredLspService calls each time will matter.

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs
Comment (ryanbrandenburg) [inline src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:49]: Curious why this couldn't apply to the context anymore?

File: src/Features/LanguageServer/Protocol/LspServices/ExportLspServiceFactoryAttribute.cs
Comment (ryanbrandenburg) [inline src/Features/LanguageServer/Protocol/LspServices/ExportLspServiceFactoryAttribute.cs:10]: Nit: In the other files you were using file-scope namespace.

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/AbstractVSTypeScriptRequestHandlerProvider.cs
Comment (sharwell) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/AbstractVSTypeScriptRequestHandlerProvider.cs:11]: ⚠️ This is problematic. Do we know if the VSTypeScript services are stateful?

File: src/Features/LanguageServer/Protocol/LspServices/LspServices.cs
Comment (ryanbrandenburg) [inline src/Features/LanguageServer/Protocol/LspServices/LspServices.cs:19]: readonly?

File: src/Features/LanguageServer/Protocol/LspServices/ExportRoslynLspServiceFactoryAttribute.cs
Comment (sharwell) [inline src/Features/LanguageServer/Protocol/LspServices/ExportRoslynLspServiceFactoryAttribute.cs:17]: ❔ Why `Roslyn` in the name of this attribute?

File: src/Features/LanguageServer/Protocol/LspServices/ExportRoslynLspServiceFactoryAttribute.cs
Comment (sharwell) [inline src/Features/LanguageServer/Protocol/LspServices/ExportRoslynLspServiceFactoryAttribute.cs:25]: 📝 Same comment about `Roslyn` in the name applies here

**Review by ryanbrandenburg — COMMENTED** [review]
Some small questions and comments but it looks good. Probably you'll want a review from someone with more Roslyn background though.

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/XamlRequestDispatcherFactory.cs
Comment (sharwell) [inline src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/XamlRequestDispatcherFactory.cs:92]: 💡 Need to enable the one-type-per-file analyzer

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/AbstractVSTypeScriptRequestHandlerProvider.cs
Comment (dibarbet) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/Api/AbstractVSTypeScriptRequestHandlerProvider.cs:11]: as of today, there aren't any (this was added recently) - there will be some coming and currently will be stateful

File: src/Features/LanguageServer/Protocol/LspServices/ExportRoslynLspServiceFactoryAttribute.cs
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/LspServices/ExportRoslynLspServiceFactoryAttribute.cs:17]: this was to differentiate between the c#/vb ones vs. Xaml vs. JS.  I can potentially rename to CSharpVisualBasicXXX though.

File: src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionResolveHandler.cs
Comment (dibarbet) [reply src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionResolveHandler.cs:73]: yeah - the cache implementation remains unchanged (must be thread safe).

File: src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsCache.cs
Comment (dibarbet) [reply src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsCache.cs:29]: To clarify - `CreateILspService` is called once per server.  Then everything in that server is handed a `LspServices` instance that holds onto the appropriate lsp services for that server.

So on each GetService call it returns the instance created for that server.  I have unit tests that verify this, an assert here would be kind of hard to write.

File: src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsHandler.cs
Comment (dibarbet) [reply src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsHandler.cs:64]: I don't think it particularly matters either way - doing it this way reduces the boiler plate a bit and the handler can become stateless.  At the end of the day it is just a lookup into a map so there isn't really much cost here. 

If I store it on the type, the handler has to become stateful which is fine, but just more code.

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:49]: It could - but the fewer ways of getting the same thing the better.  This is a mef service so we should be retrieving it via MEF.

File: src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/XamlRequestDispatcherFactory.cs
Comment (dibarbet) [reply src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/XamlRequestDispatcherFactory.cs:92]: yeah these should probably all be split out.  will update!

File: src/Features/LanguageServer/Protocol/LspServices/LspServices.cs
Comment (davidwengier) [inline src/Features/LanguageServer/Protocol/LspServices/LspServices.cs:39]: Perhaps "Any" is a better name than "NotSpecified"? Or maybe there is a doc comment that explains this, which is probably fine, but I don't think i've got to that file yet.

File: src/Features/LanguageServer/Protocol/LspServices/LspServices.cs
Comment (davidwengier) [inline src/Features/LanguageServer/Protocol/LspServices/LspServices.cs:42]: Is there a reason not to just concat this together with `servicesFromFactories` above, and do one server kind check for the whole thing? Or at least perhaps a debug.assert that ensures the base services are all for the right server kinds? Seems like this could be a future bug farm.

File: src/Features/LanguageServer/Protocol/LspServices/ExportStatelessLspServiceAttribute.cs
Comment (davidwengier) [reply src/Features/LanguageServer/Protocol/LspServices/ExportStatelessLspServiceAttribute.cs:13]: Oh.. there is stateful services..hm.. more doc comments here about what this means for disposal might be good then

**Review by davidwengier — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/LspServices/ExportStatelessLspServiceAttribute.cs
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/LspServices/ExportStatelessLspServiceAttribute.cs:13]: 👍

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptPullDiagnosticHandlerProvider.cs
Comment (dibarbet) [reply src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptPullDiagnosticHandlerProvider.cs:27]: Possibly - however not changed in this PR and don't think it is super impactful.  Can followup later.

**Comment by dibarbet** [comment]
merging before there are merge conflicts since this touches a lot of files

if there is more feedback, will take do in a followup

**Pull request merged**
